### ISSUE TITLE
Call to unexisting SpatialConvolutionMM_updateOutput

### ISSUE THREAD

**Issue by sheerun** [chuanli11/CNNMRF#9]
I get following on both cudnn (titan x) and opencl (osx, iris) backends: 

```
 /Users/sheerun/Source/CNNMRF/mylib/mrf.lua:152: attempt to call field 'SpatialConvolutionMM_updateOutput' (a nil value)
stack traceback:
    /Users/sheerun/Source/CNNMRF/mylib/mrf.lua:152: in function 'updateGradInput'
    ...sheerun/Source/torch/install/share/lua/5.1/nn/Module.lua:30: in function 'backward'
    ...run/Source/torch/install/share/lua/5.1/nn/Sequential.lua:84: in function 'backward'
    ./syn_CNNMRF_wrapper.lua:303: in function 'opfunc'
    /Users/sheerun/Source/CNNMRF/mylib/myoptimizer.lua:34: in function 'mylbfgs'
    ./syn_CNNMRF_wrapper.lua:458: in function 'main'
    ./syn_CNNMRF_wrapper.lua:500: in function 'state'
    run_syn.lua:59: in main chunk
    [C]: in function 'dofile'
    ...urce/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:145: in main chunk
    [C]: at 0x0104104bb0
```

Any idea how to fix it?

**Comment by sheerun** [comment]
Here is what `print(input.nn)` returns on os x, opencl:

```
{
  LogSigmoid_updateOutput : function: 0x067d5d18
  HardTanh_updateOutput : function: 0x0677c7d0
  Abs_updateGradInput : function: 0x0679e318
  Sqrt_updateGradInput : function: 0x06d58348
  Square_updateOutput : function: 0x06d58478
  Square_updateGradInput : function: 0x06d58580
  Abs_updateOutput : function: 0x0679e2d8
  Sigmoid_updateOutput : function: 0x06cee0f8
  Exp_updateGradInput : function: 0x0679e170
  Sqrt_updateOutput : function: 0x06bfbb10
  Exp_updateOutput : function: 0x0679e100
  Sigmoid_updateGradInput : function: 0x06cf3d28
  LogSigmoid_updateGradInput : function: 0x067d5d58
  HardTanh_updateGradInput : function: 0x0677c810
}
```

**Comment by sheerun** [comment]
For ubuntu 14.04  / cudnn machine it's following:

```
{
  TemporalConvolution_accGradParameters : function: 0x404d3370
  SpatialCrossMapLRN_updateGradInput : function: 0x41c4a610
  SpatialFullConvolution_updateOutput : function: 0x404d3578
  SpatialConvolutionLocal_updateGradInput : function: 0x404d3528
  SpatialFractionalMaxPooling_updateOutput : function: 0x406e4168
  SpatialConvolutionLocal_accGradParameters : function: 0x404d3550
  TemporalConvolution_updateGradInput : function: 0x404d3348
  SpatialBatchNormalization_backward : function: 0x404d34d8
  SpatialUpSamplingNearest_updateOutput : function: 0x404d3398
  TemporalConvolution_updateOutput : function: 0x41c4a638
  SpatialBatchNormalization_updateOutput : function: 0x404d34b0
  SpatialConvolutionLocal_updateOutput : function: 0x404d3500
  SpatialSubSampling_accGradParameters : function: 0x404d3308
  SpatialSubSampling_updateGradInput : function: 0x404d32e0
  SpatialSubSampling_updateOutput : function: 0x41c4a688
  SpatialFractionalMaxPooling_updateGradInput : function: 0x41c4a660
  SpatialMaxUnpooling_updateGradInput : function: 0x41c4a5d8
  TemporalMaxPooling_updateGradInput : function: 0x404d3488
  SpatialFullConvolution_updateGradInput : function: 0x404d35a0
  SpatialCrossMapLRN_updateOutput : function: 0x41c4a5b0
  SpatialMaxUnpooling_updateOutput : function: 0x404d35f0
  SpatialUpSamplingNearest_updateGradInput : function: 0x404d33c0
  SpatialFullConvolution_accGradParameters : function: 0x404d35c8
  TemporalMaxPooling_updateOutput : function: 0x404d3460
}
```

No `SpatialConvolutionMM_updateOutput` is defined.

**Comment by dezza73** [comment]
I have the same problem.

I'm running on ubuntu 15.10. clnn using intel opencl gpu. neural-style runs fine.

**Comment by motifMachine** [comment]
Got same problem. Could it have something to do with cudnn 4?

**Comment by SMDevlin** [comment]
I had the same issue and resolved it by updating both nn and cunn following the advice given here: https://groups.google.com/d/msg/torch7/CnkUoI6Dh1E/wZrXJaXTAwAJ

**Comment by hermitman** [comment]
Got the same error, and Resolved it as mentioned by @SMDevlin.

However, new error came out saying:

mrf.lua:152 attempt to index field 'nn (a nil value)'

Please help. I have been stucking on line 152 for a long time.....

**Comment by dezza73** [comment]
I have exactly the same issue as hermitman. I installed ubuntu from scratch with an nvidia card and using cudnn 4. neural-style which has the same requirements runs fine on two machines.

If anyone has this version working what setup/install steps do you use?

thanks very much!

**Comment by chuanli11** [comment]
@hermitman @dezza73 Have you tried re-installing torch? (Then update the dependency as suggested by SMDevlin)

**Comment by motifMachine** [comment]
I did everything. Got different error message, but still stuck on the same
line. Since I am very new to torch, debug this is kind of impossible for
me. Would really appreciate some help.

**Comment by hermitman** [comment]
@chuanli11: I followed the same steps as described by @SMDevlin. I also did a clean re-install of torch, nn, and cunn. I always stuck in line 152 for either:
1. attempt to call field 'SpatialConvolutionMM_updateOutput' (a nil value)
2. attempt to index field 'nn (a nil value)'

thanks

**Comment by nomadshark** [comment]
@chuanli11, 
the latest nn module has change its original api input.nn.SpatialConvolutionMM_updateOutput to input.THNN.SpatialConvolutionMM_updateOutput
so it will cause the "nil" error

**Comment by motifMachine** [comment]
Hi nomadshark,

Thanks for the help. I updated the code. However now I got following error.
Could you give me some further suggestions?

~/torch/install/bin/luajit: ~/torch/install/share/lua/5.1/nn/THNN.lua:1045:
bad argument #2 to 'v' (cannot convert 'table' to 'struct THCudaTensor *')

stack traceback:

[C]: in function 'v'

~/torch/install/share/lua/5.1/nn/THNN.lua:1045: in function
'SpatialConvolutionMM_updateOutput'

~/CNNMRF/CNNMRFOLD/mylib/mrf.lua:119: in function 'updateGradInput'

**Comment by hermitman** [comment]
Anyone got the program running? Please let us know what should be done to solve the issue please. Thanks~

**Comment by austingg** [comment]
I have just run the program right and haven't got you guy mentioned error. The basic style transfer environment such as torch, cuda, cudnn, is based on Justin's implementation setting. Everything seems ok. And the result seem better than neural style when the style image and content image are of the same theme.

**Comment by hermitman** [comment]
@austingg 

Glad to hear the success. Now, could you help me to reproduce your success by let me know the following:
1. your torch version
2. your cudnn version

Besides, if you have time, could you help me run a test as follows:
1. clone and install the latest Torch7.
2. install cudnn4
3. run CNNMRF.

Thanks a ton,

**Comment by austingg** [comment]
@hermitman have ever tried the original famous implementation of neural style by [Justin](https://github.com/jcjohnson/neural-style). there are detailed setting ups. you can reference.
my current cudnn version is v3.

**Comment by hermitman** [comment]
@austingg 

Yes, the program runs well on my system. I checked the code. Neural-Style do not use:

input.nn.SpatialConvolutionMM_updateOutput(self, input)

which is where everyone gets stuck. It appears that the latest NN update uses CUDNN4, which somehow broke:

input.nn

if I run 

print('input')

I got the matrix.

however, running

print('input.nn')

gives me the same error as using this program.

BTW, could you share a version of the torch, cunn, nn that you are using? I struggled hard to locate a old distribute of Torch that do not use cuDNN 4.

thanks,

**Comment by hermitman** [comment]
@austingg 

Could you send me a copy of your "[torch root]/extra/nn/SpatialConvolutionMM.lua"?

I am new to Torch and LUA, and debugging this program drove me nuts, consider that I do not understand the syntax of LUA program - -!

Thanks,

**Comment by hermitman** [comment]
Hi, All:

I can confirm the problem is with nn now, as I find a old commit from torch/nn.

In file "nn/SpatialConvolutionMM.lua", they used to have:

function SpatialConvolutionMM:updateOutput(input)
   -- backward compatibility
   if self.padding then
      self.padW = self.padding
      self.padH = self.padding
      self.padding = nil
   end
   input = makeContiguous(self, input)
   return **input.nn.SpatialConvolutionMM_updateOutput**(self, input)
end

After the update, the function becomes:

function SpatialConvolutionMM:updateOutput(input)
   self.finput = self.finput or input.new()
   self.fgradInput = self.fgradInput or input.new()
   -- backward compatibility
   if self.padding then
      self.padW = self.padding
      self.padH = self.padding
      self.padding = nil
   end
   input = makeContiguous(self, input)
   input.THNN.SpatialConvolutionMM_updateOutput(
      input:cdata(),
      self.output:cdata(),
      self.weight:cdata(),
      self.bias:cdata(),
      self.finput:cdata(),
      self.fgradInput:cdata(),
      self.kW, self.kH,
      self.dW, self.dH,
      self.padW, self.padH
   )
   return self.output
end

Thus, 

input.nn is deprecated, and cannot be used anymore. 

Due to that I have only used Torch for 3 days, I cannot solve this problem. If anyone knows what to use to bridge the gap. Please let us know.

Thanks,

**Comment by austingg** [comment]
@hermitman I am also newbee for Torch7. the environment is setting last year, when I tried neural style. So cudnn v3, cuda 6.5? besides, could u tell me  how to get torch version?

**Comment by austingg** [comment]
@hermitman And if you use gpu, maybe don't need nn.SpatialxxxxMM

**Comment by hermitman** [comment]
@austingg 
I followed this: http://torch.ch/docs/getting-started.html#_
to install torch.

After install torch, you will go to Nvidia and download cudnn4 (registration needed). Then, you can try to run the CNNMRF and use luarocks to install the missing packages, load-caffe, etc

**Comment by hermitman** [comment]
@austingg 

It appears to me that your old torch is very valuable to us now. Could you send me the entire package via dropbox or anything.... I don't think I can get CNNMRF to work with the new torch and new nn package easily.....

**Comment by hermitman** [comment]
@chuanli11 

I am modifying the code to work with new Torch, and I got confused with one thing in mrf.lua.

While initializing the mrf layer. the number of channel of the bias is set to the same as "nOutputPlane". However, the number of channel of weight is set to the same as the gpu_gpu_chunck_size_1. What is the reason for such setting? Could I modify the initialization of bias, so it has the same number of channel as weight?

thank,

**Comment by hermitman** [comment]
Problem solved. @chuanli11  please let me know if I should prepare a pull request for it.

**Comment by chuanli11** [comment]
@hermitman Thanks a lot for your effort. Please send me a pull request.

Chuan (Chris) Li

Postdoc Researcher

Mainz University

**Comment by hermitman** [comment]
@chuanli11 I just tried to do it. However, I noticed a new update has been merged and I need to pull that first. I will do it once I finish my experiments.

thanks,

**Comment by bododge** [comment]
Hello, I'm new to this project, but I'm just checking in to say I'm having the same problem as everyone else. I get stuck at- first
attempt to call field 'SpatialConvolutionMM_updateOutput' (a nil value) 

and then after fixes:
attempt to index field 'nn (a nil value)'

If you get it working, please tell us how!

**Comment by hermitman** [comment]
@bododge I am busy with something else that I don't want to risk my local working code by pulling in new update. However, I think replacing line 152 with the following should help you:

local subBias = self.bias:sub(i_start, i_end)

```
input.THNN.SpatialConvolutionMM_updateOutput(
  input:cdata(),
  self.output:cdata(),
  self.weight:cdata(),
  subBias:cdata(),
  self.finput:cdata(),
  self.fgradInput:cdata(),
  self.kW, self.kH,
  self.dW, self.dH,
  self.padW, self.padH
)
local temp = self.output
```

let me know if it works for you too.

**Comment by bododge** [comment]
@hermitman Yes that worked for me too! Thank you so much - really I am so excited about this project, THANK YOU. I'm running latest version of everything btw. I went through all of my dependency installs several times today so everything is up to date. OS X 10.10.5, Torch, CuDNN 4, cuda 7.5, latest nn, cunn, clnn, cudnn.torch. Your change looks like a good candidate for committing to main branch. Nice work & thank you for helping!

**Comment by dezza73** [comment]
@hermitman It worked with that change - thanks very much, fantastic!

**Comment by sunil-log** [comment]
@hermitman It worked for me too. Thank you!

**Comment by bododge** [comment]
@chuanli11 & @hermitman  [T.hanks for your amazing work](http://imgur.com/gallery/VVLeTmE)

**Comment by ju40268** [comment]
@chuanli11 
I still got the problem:
![image](https://cloud.githubusercontent.com/assets/8717276/13723822/cb1d0af0-e8ab-11e5-9787-e8f3672a49ad.png)
after i had changed the file with SpatialConvolutionMM:updateOutput() as followed:
![image](https://cloud.githubusercontent.com/assets/8717276/13723841/354faf04-e8ac-11e5-8230-02a6e583efeb.png)
in these two files 
./torch/install/share/lua/5.1/nn/SpatialConvolutionMM.lua
./torch/extra/nn/SpatialConvolutionMM.lua
are there some other possible bugs to fix?
thanks for help!

**Comment by mrmartin** [comment]
The fix proposed by @hermitman was included in pull request #15, which has now been merged into the repo.
Since that seems to work for everybody, this thread can be closed.

**Comment by KangolHsu** [comment]
luarocks install nn or cunn failed, So I reinstall torch ,then the error disappear
